### ISSUE TITLE
The proper way to check if the stream exists

### ISSUE THREAD

**Issue by swanden** [nats-io/nats.go#1087]
Hi!
I'm using NATS Go client https://github.com/nats-io/nats.go
I'm trying to add a stream. Before that, I'm checking if stream exists:

```
nc, _ := nats.Connect(nats.DefaultURL)
js, _ := nc.JetStream()

stream, err := js.StreamInfo(streamName)
if err != nil {
    if errors.Is(err, nats.ErrStreamNotFound) || server.IsNatsErr(err, server.JSStreamNotFoundErr) {
        return false, nil
    }

    return false, errors.Wrap(ErrStreamInfo, err.Error())
}

if stream != nil {
    return
}

if _, err := js.AddStream(&nats.StreamConfig{
    Name:     streamName,
    Subjects: []string{fmt.Sprintf("%s.*", streamName)},
}); err != nil {
    return errors.Wrap(ErrStreamAdd, err.Error())
}
```

Everything works fine, but I have an increment of the errors count in /jsz stats page. StreamInfo() returinig 404 error: stream not found.

```
{
"server_id": "...",
"now": "2022-09-16T11:51:43.41376574Z",
"config": {
    "max_memory": 1000000000,
    "max_storage": 1000000000,
    "store_dir": "nats/jetstream"
},
"memory": 0,
"storage": 0,
"accounts": 1,
"api": {
    "total": 0,
    "errors": 1
},
...
}
```

Is StreamInfo() call the right way to check if stream exists or there is a better solution?
Should the counter always be zero or is it a sign that something is going wrong and I'm not using the Go client api correctly?
Do I really need to do a check for the existence of a stream before add it and start publishing messages to it?

**Comment by piotrpio** [comment]
Hi @swanden! 

In general yes, `StreamInfo()` is the correct way to check if a stream exists. However, `AddStream()` call is idempotent, so as long as your `StreamConfig` is not changed, you don't need any additional checks before adding the stream. Just note that if a stream with the provided name exists and has different config than the one you supplied, `AddStream()` will return `nats.ErrStreamNameAlreadyInUse`.

One side note - when checking if stream exists, you can drop the check for `server.IsNatsErr(err, server.JSStreamNotFoundErr)` - go client will always return `nats.ErrStreamNotFound` if a stream does not exists.

To sum up, your code could look something like this:
```go
nc, _ := nats.Connect(nats.DefaultURL)
js, _ := nc.JetStream()

_, err := js.AddStream(&nats.StreamConfig{
	Name:     streamName,
	Subjects: []string{fmt.Sprintf("%s.*", streamName)},
})
if err != nil {
	if errors.Is(err, nats.ErrStreamNameAlreadyInUse) {
		// stream already exists and has different config
	}
	return errors.Wrap(ErrStreamAdd, err.Error())
}
```

**Comment by swanden** [comment]
Thank you!
